### DISCUSSION TITLE
How to donwload a div's content as a JPEG file?

### DISCUSSION THREAD

**Discussion by rezarahem** [vercel/next.js#47259] [Help] [answered]
### Summary

Hi 
I'm designing a business card generator, but don't know how to download the card. It's a just a simple div. Is there any way to download this div as JPEG?

**Comment by unleashit** [comment]
Interesting process.  I'm not sure if this has anything to do with Next.js but are you basically wanting a screenshot of the rendered div... or just the contents? Would need more detail on the setup. For the latter, it would depend on if you have access to the code that produces it. If you don't, you might have to scrape the page. But if you're trying to get a formatted version that looks like the rendering on the screen, you probably need a screenshot. That means you'll probably have to render it in a headless browser like puppeteer, take a screenshot and crop (can be automated). That's beyond the scope of Next.js and be careful of resolution issues. Printing requires much higher res than screen and it's best to preserve the text and other elements as vector if you can.

**Comment by SRX9** [comment]
I did that a while ago, you can use the "html2canvas" npm package. It works so quickly.  

You can refer to this simple guide, if you need more help and a small example for it: [Convert Div to Image and download](https://ayedot.com/6104/MiniBlog/How-to-Convert--Download-HTML-or-divs-Content-as-Image-in-any-JavaScript-Framework)

``` js
import React, { useRef, useState } from "react";
import html2canvas from "html2canvas";

const DivToImage = () => {
  const exportDivImageRef: any = useRef();
  const [downloading, setDownloading] = useState<boolean>(false);

  const exportAsImage = async (element, imageFileName) => {
    setDownloading(true);
    try {
      const html: any = document.getElementsByTagName("html")[0];
      const body: any = document.getElementsByTagName("body")[0];
      let htmlWidth = html.clientWidth;
      let bodyWidth = body.clientWidth;
      const newWidth = element.scrollWidth - element.clientWidth;
      if (newWidth > element.clientWidth) {
        htmlWidth += newWidth;
        bodyWidth += newWidth;
      }
      html.style.width = htmlWidth + "px";
      body.style.width = bodyWidth + "px";
      const canvas = await html2canvas(element, {
        allowTaint: true,
        useCORS: true,
      });
      const image = canvas.toDataURL("image/png", 1.0);
      downloadImage(image, imageFileName);
      html.style.width = null;
      body.style.width = null;
    } catch (error) {
      setDownloading(false);
    }
  };

  const downloadImage = (blob, fileName) => {
    try {
      const fakeLink: any = window.document.createElement("a");
      fakeLink.style = "display:none;";
      fakeLink.download = fileName;

      fakeLink.href = blob;

      document.body.appendChild(fakeLink);
      fakeLink.click();
      document.body.removeChild(fakeLink);

      fakeLink.remove();
      setDownloading(false);
      setDownloading(false);
    } catch (error) {
      setDownloading(false);
    }
  };

  return (
    <>
      <div ref={exportRef}>
        <h4>Hello world on Image!</h4>
      </div>
      <button onClick={downloadImage}>Download Card Image</button>
    </>
  );
};
```

**Comment by icyJoseph** [comment, marked as answer]
Not sure I am being helpful by pasting code here, but anyway:

```tsx
import { useRef } from "react";
import Image from "next/image";

// leaving these here you know you can tweak stuff further
const options = {
  allowTaint: true,
  useCORS: true,
  backgroundColor: "rgba(0,0,0,0)",
  removeContainer: true,
};

export default function Home() {
  const cardRef = useRef<HTMLElement>(null);

  const prepareURL = async () => {
    const cardElement = cardRef.current;

    if (!cardElement) return;

    try {
      // lazy load this package
      const html2canvas = await import(
        /* webpackPrefetch: true */ "html2canvas"
      );

      const result = await html2canvas.default(cardElement, options);

      const asURL = result.toDataURL("image/jpeg");
      // as far as I know this is a quick and dirty solution
      const anchor = document.createElement("a");
      anchor.href = asURL;
      anchor.download = "your-card.jpeg";
      anchor.click();
      anchor.remove();
     // maybe this part should set state with `setURLData(asURL)`
     // and when that's set to something you show the download button 
     // which has `href=URLData`, so that people can click on it
    } catch (reason) {
      console.log(reason);
    }
  };

  return (
    <div>
      <button onClick={prepareURL}>Download</button>

      <article ref={cardRef}>
        <h2>Mew</h2>

        <p># 151</p>
        {/* had to add `images: { domains: ["raw.githubusercontent.com"], }` to my next.config.js */}
        <Image
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/151.png"
          alt="mew"
          width="180"
          height="180"
        />
      </article>
    </div>
  );
}
```

Make sure your card is fully styled, and isolated. Otherwise, if the card relies on the body having some background color, or some cascading CSS, you'll not get the correct image.

**Reply by rezarahem** [reply]
thanks dude, didn't get even one error. 🙏🏻

**Comment by unleashit** [comment]
Oh yeah, don't know why I didn't think of canvas. I once made a mini Tattoo designer (for printing) with http://fabricjs.com/. It's an old library and would need quite a bit more logic, but with it you can end up with something of true print quality. I checked out HTMLtoCanvas and while neat and might work just fine for your use case, but you're probably going to end up with low resolution images and text, which will look fine on the screen, but blurry if printed. Since it's canvas, you should theoretically at least be able to output as an SVG which could help with the text, and maybe even get there with images if you set a decent min size for the user added images (you can temp downsize with css in the div for display reasons).

**Reply by rezarahem** [reply]
thanks bro. actually that's one of my concerns.
